### ISSUE TITLE
Quick hack for slow startup when using python

### ISSUE THREAD

**Issue by critiqjo** [neovim/neovim#2094]
Starting neovim was taking ~0.5 seconds for me (python-client installed), while vim was only taking ~0.09s
Digging in, I found a quick hack -- change `/usr/share/nvim/runtime/autoload/provider/python.vim` like this:
```
if exists('s:loaded_python_provider') || &cp
  finish
endif
let s:loaded_python_provider = 1
let s:plugin_path = expand('<sfile>:p:h').'/script_host.py'

call remote#host#RegisterClone('legacy-python-provider', 'python')
call remote#host#RegisterPlugin('legacy-python-provider', s:plugin_path, [])

let s:rpcrequest = function('rpcrequest')

function! provider#python#Call(method, args)
  if !exists('s:host')
    let s:host = remote#host#Require('legacy-python-provider')
  endif
  return call(s:rpcrequest, insert(insert(a:args, 'python_'.a:method), s:host))
endfunction
```
`:py` scripts still works fine. And startup time reduced to ~0.09s! Yay!
I hope a better, more general solution comes out soon.
Thank you for the great work guys!

**Caveat:** Unless you have [python-client](https://github.com/neovim/python-client) installed, do not apply this hack.

Current version: 6b7ece6cc89aa35f99d6bd3c078114a7a8a2daa1
OS: Arch Linux

**Comment by nanomad** [comment]
I can confirm this patch fixes slow start-up times for me too. I've also re-added the try/catch block just in case

**Comment by Pyrohh** [comment]
This is indeed faster when I don't have any python plugins (i.e. I only have the python plugin host installed), but has a very big issue for me.

Both of these examples assume nvim is installed along with UltiSnips, but the nvim python client isn't installed

**nvim master:**

Start nvim, encounter the following error:
```
No neovim module found for Python 2.7.9. Try installing it with 'pip install neovim' or see ':help nvim-python'.
UltiSnips requires py >= 2.6 or any py3
Press ENTER or type command to continue
```
After pressing the enter key, everything works as you would expect (i.e., no ultisnips, but everything else works fine).

**your version:**

Start nvim, encounter the following error:
```
Error detected while processing function UltiSnips#bootstrap#Bootstrap..provider#python#Call..remote#host#Require..<SNR>4_RequirePythonHost:
line   55:
E605: Exception not caught: No neovim module found for Python 2.7.9. Try installing it with 'pip install neovim' or see ':help nvim-python'.
Error detected while processing function UltiSnips#bootstrap#Bootstrap..provider#python#Call..remote#host#Require:
line    6:
E171: Missing :endif
Error detected while processing function UltiSnips#bootstrap#Bootstrap..provider#python#Call:
line    2:
E171: Missing :endif
```

UltiSnips then throws a bunch of errors every time I move my cursor.

**Comment by critechar** [comment]
@Pyrohh That is indeed true. Which is why I requested a more general solution. This "hack" makes nvim (and plugins) think that it has python support regardless of reality. So unless you have python-client installed, do not apply this hack.

**Comment by Pyrohh** [comment]
I see, thanks for clarifying how it works.

**Comment by Shougo** [comment]
I get the problem.
But it is difficlut to fix it.
You should use lazy loading by plugin managers.

**Comment by Pyrohh** [comment]
@Shougo I don't think I understand what you mean. I get this startup delay with or without Python plugins.

**Comment by Shougo** [comment]
Really?  I confirmed it is loaded when using `has('python')`.

**Comment by Pyrohh** [comment]
There might be a lapse in communication here :-)

> I confirmed it is loaded when using has('python').

The same happens to me, but what I was trying to say earlier is that I don't think a lazy-loading plugin manager will fix the issue.

For me, nvim tries to load the Python client if it's available (which is always), which makes nvim take longer to startup. It does this regardless of if I'm running Python plugins or not, so with this said then I don't think lazy loading will make a difference.

**Comment by Shougo** [comment]
I think you have `:has('python')` check in your .vimrc.

I tested `:has('python')` check and it takes 0.5s time.
Yes, it is too slow.

I will improve the time.

Unfortunatelly, this hack breaks plugin compatibility.
So it is not default.  But I can add the option for it.

**Comment by Pyrohh** [comment]
> Unfortunatelly, this hack breaks plugin compatibility.
> So it is not default. But I can add the option for it.

I don't think we should have an option for this, given how much it breaks when it does. That is unless you mean you plan to make if fail more quietly.

**Comment by Shougo** [comment]
> I don't think we should have an option for this, given how much it breaks when it does.

If `has('python')` is 1, neovim must work if_python.
But the hack skips the check after executed `:py`.
I think it breaks the compatilibility.

**Comment by blueyed** [comment]
I think this is provided by #2208 now, or at least should be followed up there..

**Comment by Pyrohh** [comment]
> I think this is provided by #2208 now, or at least should be followed up there..

I think your second part was right, as my startup times are still slow on that branch. By the way, thanks for helping out.

**Comment by blueyed** [comment]
@Pyrohh 
I am using the following with #2208.

It disables the (autoloaded) Python 3 host (which should result in
`has("python3") == 0`) and adds a message for debugging purposes.

I've also defined `python_host_prog` explicitly to improve/skip the detection.

With this (and #2208) in general, the provider only gets used when it's being requested.  But this might get triggered by `has()` and there are plugins that prefer Python 3 (UltiSnips), while most others ask for Python 2 first.  This results in both hosts being started then.

    diff --git a/runtime/autoload/provider/python3.vim b/runtime/autoload/provider/python3.vim
    index 5eb06a6..359640e 100644
    --- a/runtime/autoload/provider/python3.vim
    +++ b/runtime/autoload/provider/python3.vim
    @@ -4,7 +4,7 @@
     " Associating the plugin with the Python3 host is the first step because
     " plugins will be passed as command-line arguments
     
    -if exists('s:loaded_python3_provider')
    +if 1 || exists('s:loaded_python3_provider')
       finish
     endif
     
    diff --git a/runtime/autoload/remote/host.vim b/runtime/autoload/remote/host.vim
    index 3ae6a58..1e24014 100644
    --- a/runtime/autoload/remote/host.vim
    +++ b/runtime/autoload/remote/host.vim
    @@ -206,6 +206,9 @@ function! s:RequirePythonHost(name)
       try
         let channel_id = rpcstart(python_host_prog, args)
         if rpcrequest(channel_id, 'poll') == 'ok'
    +      if python_host_prog != "…/pyenv/2.7.9/bin/python2"
    +        echom "Started also:" python_host_prog
    +      endif
           return channel_id
         endif
       catch

**Comment by Shougo** [comment]
> I think this is provided by #2208 now, or at least should be followed up there..

Yes, I must improve the initialization #2208.

**Comment by Shougo** [comment]
I improved the initialization in https://github.com/neovim/neovim/pull/2208.

**Comment by critiqjo** [comment]
Fixed by #2208...
